packages/core/src/hooks.ts

```diff
@@ -230,15 +232,16 @@ const onBeforeCompile = ({
             compilerDone = true;
             doneCompilers++;
           }
-          // ensure only last compiler done will trigger beforeCompile hook
-          // avoid other compiler done triggers when executing async beforeEnvironmentCompiler hook
-          const lastCompilerDone = doneCompilers === compilers.length;
-
-          await beforeEnvironmentCompiler(index);
+          const isFirstTrigger = doneCompilers === 1;
 
-          if (lastCompilerDone) {
-            await beforeCompile?.();
+          if (isFirstTrigger) {
```

Comment:
It seems we do not need a `isFirstTrigger` variable:

```suggestion
          if (!waitBeforeCompileDone) {
```

Reply:
`doneCompilers` is not reliable as it can decrease in some cases.